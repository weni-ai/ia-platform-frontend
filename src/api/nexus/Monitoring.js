import request from '@/api/nexusaiRequest';
import { cleanParams } from '@/utils/http';

export const Monitoring = {
  messages: {
    async list({
      projectUuid,
      page,
      pageInterval,
      tag,
      text,
      started_day,
      ended_day,
    }) {
      const params = cleanParams({
        tag,
        text,
        started_day,
        ended_day,
        page_size: pageInterval,
        page,
      });
      const {
        data: { results, count },
      } = await request.$http.get(`api/${projectUuid}/message_history/`, {
        params,
      });

      return {
        count,
        data: results.map(
          ({ id, created_at, message_text, tag, classification }) => ({
            created_at,
            id,
            text: message_text,
            tag,
            action_name: classification,
          }),
        ),
      };
    },

    async history({ projectUuid, id }) {
      // const {
      //   data: { results },
      // } = await request.$http.get(
      //   `api/${projectUuid}/message_history/?uuid=${id}`,
      // );

      return {
        uuid: 'uuid',
        contact_urn: 'whatsapp:999999999',
        text: 'como envio meu curr√≠culo??',
        llm_response:
          'Desculpe, mas essa pergunta est√° fora do meu escopo de conhecimento. Posso te ajudar com outras d√∫vidas relacionadas a lhamas? ü¶ô',
        llm_response_status: 'failed',
      };
    },

    async performance({ projectUuid, started_day, ended_day }) {
      const params = cleanParams({
        started_day,
        ended_day,
      });
      const { data } = await request.$http.get(
        `api/${projectUuid}/tags-analytics/`,
        {
          params,
        },
      );

      return data;
    },
  },
};
