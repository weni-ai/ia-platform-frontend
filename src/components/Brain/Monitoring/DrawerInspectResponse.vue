<template>
  <UnnnicDrawer
    :modelValue="modelValue"
    :title="$t('router.monitoring.inspect_response.title')"
    :description="
      $t(
        'router.monitoring.inspect_response.intelligent_agent_response_details',
      )
    "
    @close="close"
  >
    <template #content>
      <section class="inspect-response">
        <section class="inspect-response__contact">
          <UnnnicIntelligenceText
            color="neutral-cloudy"
            family="secondary"
            size="body-gt"
            tag="p"
          >
            {{ $t('router.monitoring.inspect_response.contact_message') }}:
          </UnnnicIntelligenceText>
          <p class="contact__text">“{{ inspectionData.text }}”</p>
        </section>
        <section class="inspect-response__agent">
          <UnnnicIntelligenceText
            color="neutral-cloudy"
            family="secondary"
            size="body-gt"
            tag="p"
          >
            {{ $t('router.monitoring.inspect_response.agent_response') }}:
          </UnnnicIntelligenceText>
          <p
            v-if="inspectionData.llm.status !== 'success'"
            class="agent__text"
          >
            “{{ inspectionData.llm.response }}”
          </p>
          <section v-else>
            <template
              v-for="fragment of groundednessFragments"
              :key="fragment.sentence"
            >
              <p
                :class="[
                  'agent__text',
                  'agent__text--underlined',
                  `agent__text--underlined-${getReliabilityLevel(fragment)}`,
                ]"
              >
                {{ fragment.sentence }}
              </p>
            </template>
          </section>
        </section>
      </section>
    </template>
  </UnnnicDrawer>
</template>

<script setup>
import { computed } from 'vue';

const props = defineProps({
  modelValue: {
    type: Boolean,
    default: false,
  },

  inspectionData: {
    type: Object,
    required: true,
  },
});

const emit = defineEmits(['update:model-value']);

const groundednessFragments = computed(() => {
  const data = props.inspectionData.groundedness.map((fragment) => ({
    ...fragment, // Required to create copies of the fragments and not update the prop
  }));

  if (data.length > 0) {
    const firstFragment = data[0];
    const lastFragment = data.at(-1);

    firstFragment.sentence = `“${firstFragment.sentence}`;
    lastFragment.sentence = `${lastFragment.sentence}”`;
  }

  return data;
});

function close() {
  emit('update:model-value', false);
}

function getReliabilityLevel(item) {
  const { score } = item;

  if (score === 0) return 'unreliable';
  if (score <= 33) return 'slightly-reliable';
  if (score <= 66) return 'moderately-reliable';
  if (score <= 100) return 'highly-reliable';
  return 'unreliable';
}
</script>

<style lang="scss" scoped>
.inspect-response {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-md;

  &__contact,
  &__agent {
    display: grid;
    gap: $unnnic-spacing-nano;

    .contact__text,
    .agent__text {
      color: $unnnic-color-neutral-darkest;
      font-size: $unnnic-font-size-body-lg;
    }
  }

  &__agent {
    .agent__text {
      display: inline;

      line-height: $unnnic-font-size-body-lg + $unnnic-line-height-md;

      &--underlined {
        margin-right: $unnnic-spacing-nano;

        text-decoration: underline;
        text-decoration-color: $unnnic-color-neutral-cleanest;
        text-decoration-thickness: 3px;
        text-underline-offset: 4px;

        &-slightly-reliable {
          text-decoration-color: $unnnic-color-weni-300;
        }
        &-moderately-reliable {
          text-decoration-color: $unnnic-color-weni-600;
        }
        &-highly-reliable {
          text-decoration-color: $unnnic-color-weni-800;
        }
      }
    }
  }
}
</style>
