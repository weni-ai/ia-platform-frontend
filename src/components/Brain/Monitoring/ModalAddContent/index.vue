<template>
  <UnnnicModalDialog
    :modelValue="modelValue"
    size="lg"
    showCloseIcon
    :title="$t('router.monitoring.improve_response.add_new_content')"
    :secondaryButtonProps="{
      text: $t('cancel'),
    }"
    :primaryButtonProps="{
      text: $t('finish'),
    }"
    @update:model-value="$emit('update:modelValue', $event)"
    @secondary-button-click="$emit('update:modelValue', false)"
  >
    <UnnnicTab
      :tabs="contentTabs.map((tab) => tab.page)"
      :activeTab="activeTab"
      @change="onTabChange"
    >
      <template
        v-for="tab in contentTabs"
        :key="tab.page"
        #[`tab-head-${tab.page}`]
      >
        {{ $t(`content_bases.tabs.${tab.title}`) }}
      </template>
    </UnnnicTab>

    <section class="modal-add-content__content">
      <AddFilesContent v-show="activeTab === 'files'" />

      <AddSitesContent v-show="activeTab === 'sites'" />

      <textarea
        v-show="activeTab === 'text'"
        class="content__text"
        :placeholder="$t('content_bases.write_content_placeholder')"
        cols="30"
        rows="10"
        @input="$emit('update:modelValue', $event.target.value)"
      />
    </section>
  </UnnnicModalDialog>
</template>

<script setup>
import { ref } from 'vue';

import AddFilesContent from './AddFilesContent.vue';
import AddSitesContent from './AddSitesContent.vue';

const props = defineProps({
  modelValue: {
    type: Boolean,
    default: false,
  },
});

const emit = defineEmits(['close', 'update:modelValue']);

const contentTabs = ref([
  { title: 'files', page: 'files' },
  { title: 'sites', page: 'sites' },
  { title: 'text', page: 'text' },
]);
const activeTab = ref(contentTabs.value[0].page);
const onTabChange = (newTab) => {
  activeTab.value = newTab;
};
</script>

<style scoped lang="scss">
.modal-add-content {
  &__content {
    border-radius: $unnnic-border-radius-sm;
    border: $unnnic-border-width-thinner solid $unnnic-color-neutral-cleanest;

    padding: $unnnic-spacing-sm;

    min-height: 30vh;

    display: grid;
    align-items: center;

    .content__text {
      outline: none;
      resize: none;
      border: none;

      width: 100%;
      height: 100%;

      color: $unnnic-color-neutral-dark;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-gt;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
      font-weight: $unnnic-font-weight-regular;

      &::placeholder {
        color: $unnnic-color-neutral-cloudy;
        font-family: $unnnic-font-family-secondary;
        font-size: $unnnic-font-size-body-gt;
        line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
        font-weight: $unnnic-font-weight-regular;
      }
    }
  }
}
</style>
