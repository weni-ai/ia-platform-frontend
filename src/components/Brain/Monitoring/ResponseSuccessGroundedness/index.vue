<template>
  <section
    class="response-success-groundedness"
    data-testid="response-success-groundedness"
  >
    <TextSentences
      :fragments="groundednessFragments"
      :getReliabilityLevel="getReliabilityLevel"
    />
    <GroundednessSources
      :fragments="groundednessFragments"
      :getReliabilityLevel="getReliabilityLevel"
    />
  </section>
</template>

<script setup>
import { computed } from 'vue';

import TextSentences from './TextSentences.vue';
import GroundednessSources from './GroundednessSources.vue';

const props = defineProps({
  inspectionData: {
    type: Object,
    required: true,
  },
});

const groundednessFragments = computed(() => {
  const data = props.inspectionData.groundedness.map((fragment) => ({
    ...fragment,
  }));

  if (data.length > 0) {
    const firstFragment = data[0];
    const lastFragment = data.at(-1);

    firstFragment.sentence = `“${firstFragment.sentence}`;
    lastFragment.sentence = `${lastFragment.sentence}”`;
  }

  return data;
});

function getReliabilityLevel(item) {
  const { score } = item;

  if (score === 0) return 'unreliable';
  if (score <= 33) return 'slightly-reliable';
  if (score <= 66) return 'moderately-reliable';
  if (score <= 100) return 'highly-reliable';
  return 'unreliable';
}
</script>

<style scoped lang="scss">
.response-success-groundedness {
  display: grid;
  gap: $unnnic-spacing-sm;
}
</style>
