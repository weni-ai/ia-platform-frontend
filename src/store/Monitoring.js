import { defineStore } from 'pinia';
import { computed, reactive } from 'vue';
import { useRoute } from 'vue-router';

import nexusaiAPI from '@/api/nexusaiAPI.js';
import globalStore from '.';
import i18n from '@/utils/plugins/i18n';

export const useMonitoringStore = defineStore('monitoring', () => {
  const connectProjectUuid = computed(
    () => globalStore.state.Auth.connectProjectUuid,
  );

  const route = useRoute();

  const messages = reactive({
    status: null,
    data: [],
    count: 0,
    newMessages: [],
    performance: {
      status: null,
      action: 0,
      success: 0,
      failed: 0,
    },
    inspectedAnswer: {},
  });

  async function loadMessages({ page, pageInterval, tag, text }) {
    const { started_day, ended_day } = route.query;
    const currentNewMessages = [...messages.newMessages];

    try {
      messages.status = 'loading';
      messages.newMessages = [];

      const { data, count } = await nexusaiAPI.router.monitoring.messages.list({
        projectUuid: connectProjectUuid.value,
        page,
        pageInterval,
        tag,
        text,
        started_day,
        ended_day,
      });

      messages.data = data;
      messages.count = count;
      messages.status = 'complete';
    } catch (error) {
      messages.status = 'error';
      messages.newMessages = currentNewMessages;
    }
  }

  async function loadMessageContext({ id }) {
    messages.inspectedAnswer.context = {};
    const setStatus = (status) =>
      (messages.inspectedAnswer.context.status = status);

    try {
      setStatus('loading');

      const response =
        await nexusaiAPI.router.monitoring.messages.getMessageContext({
          projectUuid: connectProjectUuid.value,
          id: id,
        });

      messages.inspectedAnswer.context = response?.map(
        ({ id, llm_response, tag }) => ({
          id,
          llm: { response: llm_response, status: tag },
        }),
      );

      setStatus('complete');
    } catch (error) {
      setStatus('error');

      globalStore.state.alert = {
        type: 'error',
        text: i18n.global.t(
          'router.monitoring.error_loading_previous_messages',
        ),
      };
    }
  }

  async function loadMessageDetails({ id }) {
    try {
      messages.inspectedAnswer.status = 'loading';

      const response = await nexusaiAPI.router.monitoring.messages.detail({
        projectUuid: connectProjectUuid.value,
        id,
      });

      const {
        uuid,
        text,
        status,
        actions_started,
        actions_uuid,
        actions_type,
        llm_response,
        is_approved,
        contact_urn,
        groundedness,
      } = response;

      const llmStatusMap = {
        s: 'success',
        f: 'failed',
      };

      messages.inspectedAnswer = {
        id,
        uuid,
        text,
        action: actions_started
          ? {
              name: actions_type,
              uuid: actions_uuid,
            }
          : null,
        llm: {
          response: llm_response,
          status: actions_started
            ? 'action'
            : llmStatusMap[status.toLowerCase()],
        },
        contact_urn,
        is_approved,
        groundedness: groundedness?.map((item) => ({
          ...item,
          score: Number(item.score) * 10,
        })), // To turn into a hundred

        status: 'complete',
      };
    } catch (error) {
      messages.inspectedAnswer.status = 'error';
      globalStore.state.alert = {
        type: 'error',
        text: i18n.global.t('router.monitoring.error_loading_messages'),
      };
    }
  }

  async function loadMessagesPerformance() {
    const { started_day, ended_day } = route.query;
    try {
      messages.performance.status = 'loading';

      const response = await nexusaiAPI.router.monitoring.messages.performance({
        projectUuid: connectProjectUuid.value,
        started_day,
        ended_day,
      });

      messages.performance = {
        status: 'complete',
        action: response?.action_percentage || 0,
        success: response?.succeed_percentage || 0,
        failed: response?.failed_percentage || 0,
      };
    } catch (error) {
      messages.performance.status = 'error';
    }
  }

  async function rateAnswer({ is_approved }) {
    try {
      const response = await nexusaiAPI.router.monitoring.messages.rateAnswer({
        projectUuid: connectProjectUuid.value,
        id: messages.inspectedAnswer.id,
        is_approved,
      });

      messages.inspectedAnswer.is_approved = response.is_approved;
    } catch (error) {
      throw new Error(error);
    }
  }

  function createNewMessage({ message }) {
    messages.newMessages.push(message);
  }

  return {
    messages,
    loadMessages,
    loadMessageContext,
    loadMessageDetails,
    loadMessagesPerformance,
    rateAnswer,
    createNewMessage,
  };
});
