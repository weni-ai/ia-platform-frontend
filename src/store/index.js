import { createStore } from 'vuex';

import Auth from './auth';
import User from './user';
import Repository from './repository';
import Category from './category';
import CachedFetch from './cached-fetch';
import News from './news';
import Integration from './integration';
import Brain from './brain';

export const store = createStore({
  state() {
    return {
      obstructiveError: null,

      alert: null,

      modalWarn: null,

      router: {
        intelligenceUuid: null,
        contentBaseUuid: null,
      },
    };
  },

  modules: {
    Auth,
    User,
    Repository,
    Category,
    CachedFetch,
    News,
    Integration,
    Brain,
  },
});

store.dispatch('retriveAuthToken');

export default store;
