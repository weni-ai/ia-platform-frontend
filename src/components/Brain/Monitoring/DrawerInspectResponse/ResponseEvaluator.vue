<template>
  <section class="response-evaluator">
    <UnnnicDivider
      data-testid="divider"
      class="response-evaluator__divider"
      ySpacing="md"
    />

    <UnnnicIntelligenceText
      data-testid="rate-the-answer"
      class="response-evaluator__rate-the-answer"
      color="neutral-dark"
      family="secondary"
      size="body-gt"
      tag="p"
    >
      {{ $t('router.monitoring.inspect_response.rate_the_answer') }}:
    </UnnnicIntelligenceText>

    <section class="response-evaluator__actions">
      <button
        data-testid="like-button"
        :class="['actions__button', { 'actions__button--bold': isApproved }]"
        @click="setIsApproved(true)"
      >
        <UnnnicIcon
          icon="thumb_up"
          scheme="neutral-cloudy"
          size="sm"
          :filled="isApproved"
        />{{ $t('router.monitoring.inspect_response.i_liked') }}
      </button>
      <button
        data-testid="dislike-button"
        :class="[
          'actions__button',
          { 'actions__button--bold': isApproved === false },
        ]"
        @click="setIsApproved(false)"
      >
        <UnnnicIcon
          icon="thumb_down"
          scheme="neutral-cloudy"
          size="sm"
          :filled="isApproved === false"
        />{{ $t('router.monitoring.inspect_response.i_didnt_like') }}
      </button>
    </section>

    <ImproveResponse
      v-if="isApproved === false"
      class="response-evaluator__improve-response"
      type="action"
    />
  </section>
</template>

<script setup>
import { ref } from 'vue';
import { useMonitoringStore } from '@/store/Monitoring';

import ImproveResponse from './ImproveResponse.vue';

const props = defineProps({
  isApproved: {
    type: [Boolean, null],
    default: null,
  },
});

const monitoringStore = useMonitoringStore();

const isApproved = ref(props.isApproved);

async function setIsApproved(boolean) {
  const oldIsApproved = isApproved.value;
  isApproved.value = boolean;

  try {
    await monitoringStore.rateAnswer({ is_approved: boolean });
  } catch (error) {
    console.log('error', error);

    isApproved.value = oldIsApproved;
  }
}
</script>

<style lang="scss" scoped>
.response-evaluator {
  display: flex;
  flex-direction: column;

  .response-evaluator__divider {
    margin-top: 0;
  }

  &__rate-the-answer {
    margin-bottom: $unnnic-spacing-xs;
  }

  &__actions {
    display: flex;
    gap: $unnnic-spacing-sm;

    .actions__button {
      padding: 0;
      border: none;
      background: none;

      cursor: pointer;

      display: flex;
      align-items: center;
      justify-content: center;
      gap: $unnnic-spacing-nano;

      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-md;
      line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;

      &--bold {
        font-weight: $unnnic-font-weight-bold;
      }
    }
  }

  &__improve-response {
    margin-top: $unnnic-spacing-sm;
  }
}
</style>
