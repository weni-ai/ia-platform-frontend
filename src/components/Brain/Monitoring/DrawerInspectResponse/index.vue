<template>
  <UnnnicDrawer
    data-testid="drawer"
    :modelValue="modelValue"
    :title="$t('router.monitoring.inspect_response.title')"
    :description="
      $t(
        'router.monitoring.inspect_response.intelligent_agent_response_details',
      )
    "
    @close="close"
  >
    <template #content>
      <section class="inspect-response">
        <ContactMessage :text="inspectionData.text" />
        <ActionDetails
          v-if="llmStatus === 'action'"
          :action="inspectionData.action"
        />
        <AgentResponse
          v-if="llmStatus !== 'action'"
          :status="llmStatus"
          :response="inspectionData.llm.response"
          :inspectionData="inspectionData"
        />
        <ResponseEvaluator
          v-if="llmStatus !== 'failed'"
          :isApproved="inspectionData.is_approved"
        />
      </section>
    </template>
  </UnnnicDrawer>
</template>

<script setup>
import { computed } from 'vue';

import ContactMessage from './ContactMessage.vue';
import ActionDetails from './ActionDetails.vue';
import AgentResponse from './AgentResponse.vue';
import ResponseEvaluator from './ResponseEvaluator.vue';

const props = defineProps({
  modelValue: {
    type: Boolean,
    default: false,
  },
  inspectionData: {
    type: Object,
    required: true,
  },
});

const emit = defineEmits(['update:model-value']);

const llmStatus = computed(() => props.inspectionData.llm.status);

function close() {
  emit('update:model-value', false);
}
</script>

<style lang="scss" scoped>
.inspect-response {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-md;

  &__contact,
  &__agent {
    display: grid;
    gap: $unnnic-spacing-nano;

    .contact__text,
    .agent__text {
      color: $unnnic-color-neutral-darkest;
      font-size: $unnnic-font-size-body-lg;
    }
  }

  &__agent .agent__text {
    display: inline;
    line-height: $unnnic-font-size-body-lg + $unnnic-line-height-md;
  }
}
</style>
