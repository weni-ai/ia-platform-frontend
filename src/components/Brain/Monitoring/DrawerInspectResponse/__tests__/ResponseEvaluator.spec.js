import { shallowMount } from '@vue/test-utils';
import { describe, it, expect, beforeEach, vi, afterEach } from 'vitest';

import { useMonitoringStore } from '@/store/Monitoring';

import ResponseEvaluator from '../ResponseEvaluator.vue';
import i18n from '@/utils/plugins/i18n';

const rateAnswerMock = vi.fn();

vi.mock('@/store/Monitoring', () => ({
  useMonitoringStore: vi.fn(() => ({
    rateAnswer: rateAnswerMock,
  })),
}));

describe('ResponseEvaluator.vue', () => {
  let wrapper;
  let monitoringStoreMock;
  let setIsApprovedSpy;

  const mountWrapper = (props = {}) => {
    monitoringStoreMock = useMonitoringStore();

    return shallowMount(ResponseEvaluator, {
      props: {
        isApproved: null,
        ...props,
      },
    });
  };

  const contextT = (key) =>
    i18n.global.t(`router.monitoring.inspect_response.${key}`);

  beforeEach(() => {
    wrapper = mountWrapper();

    setIsApprovedSpy = vi.spyOn(wrapper.vm, 'setIsApproved');
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders the divider', () => {
    const divider = wrapper.findComponent('[data-testid="divider"]');

    expect(divider.exists()).toBe(true);
  });

  it('should renders the rate-the-answer text with correct props', () => {
    const rateText = wrapper.findComponent('[data-testid="rate-the-answer"]');
    expect(rateText.exists()).toBe(true);
    expect(rateText.props()).toMatchObject({
      color: 'neutral-dark',
      family: 'secondary',
      size: 'body-gt',
      tag: 'p',
    });
  });

  it('should renders the "I liked" button with correct text', () => {
    const likeButton = wrapper.find('[data-testid="like-button"]');
    expect(likeButton.exists()).toBe(true);
    expect(likeButton.text()).toContain(contextT('i_liked'));
  });

  it('should renders the "I didnâ€™t like" button with correct text', () => {
    const dislikeButton = wrapper.find('[data-testid="dislike-button"]');
    expect(dislikeButton.exists()).toBe(true);
    expect(dislikeButton.text()).toContain(contextT('i_didnt_like'));
  });

  it('should calls setIsApproved(true) when the like button is clicked', async () => {
    const likeButton = wrapper.find('[data-testid="like-button"]');
    await likeButton.trigger('click');
    expect(setIsApprovedSpy).toHaveBeenCalledWith(true);
  });

  it('should calls setIsApproved(false) when the dislike button is clicked', async () => {
    const dislikeButton = wrapper.find('[data-testid="dislike-button"]');
    await dislikeButton.trigger('click');
    expect(setIsApprovedSpy).toHaveBeenCalledWith(false);
  });

  it('should applies the bold class to the like button when isApproved is true', async () => {
    wrapper = mountWrapper({ isApproved: true });
    const likeButton = wrapper.find('[data-testid="like-button"]');
    expect(likeButton.classes()).toContain('actions__button--bold');
  });

  it('should applies the bold class to the dislike button when isApproved is false', async () => {
    wrapper = mountWrapper({ isApproved: false });
    const dislikeButton = wrapper.find('[data-testid="dislike-button"]');
    expect(dislikeButton.classes()).toContain('actions__button--bold');
  });

  describe('setIsApproved function', () => {
    it('should call rateAnswer with the correct value', async () => {
      await wrapper.vm.setIsApproved(true);

      expect(rateAnswerMock).toHaveBeenCalledWith({
        is_approved: true,
      });
      expect(wrapper.vm.isApproved).toBe(true);
    });

    it('should update isApproved to the new value', async () => {
      await wrapper.vm.setIsApproved(false);

      expect(wrapper.vm.isApproved).toBe(false);
    });

    describe('rateAnswer fail behaviors', () => {
      const consoleSpy = vi.spyOn(console, 'log').mockImplementation(() => {});

      beforeEach(() => {
        monitoringStoreMock.rateAnswer.mockRejectedValue(
          new Error('Store Error'),
        );
      });

      afterEach(() => {
        vi.clearAllMocks();
      });

      it('should restore the previous value of isApproved if rateAnswer fails', async () => {
        await wrapper.vm.setIsApproved(false);

        expect(wrapper.vm.isApproved).toBe(null);
      });

      it('should log an error if rateAnswer fails', async () => {
        await wrapper.vm.setIsApproved(false);

        expect(consoleSpy).toHaveBeenCalledWith('error', expect.any(Error));
      });
    });
  });
});
