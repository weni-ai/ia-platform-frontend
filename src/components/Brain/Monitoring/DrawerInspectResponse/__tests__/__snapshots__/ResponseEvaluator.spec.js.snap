// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResponseEvaluator.vue > matches the snapshot 1`] = `
"<section data-v-4cf8372f="" class="response-evaluator">
  <unnnic-divider-stub data-v-4cf8372f="" yspacing="md" scheme="neutral-soft" data-testid="divider" class="response-evaluator__divider"></unnnic-divider-stub>
  <unnnic-intelligence-text-stub data-v-4cf8372f="" size="body-gt" tag="p" family="secondary" lineheight="md" weight="regular" color="neutral-dark" data-testid="rate-the-answer" class="response-evaluator__rate-the-answer"></unnnic-intelligence-text-stub>
  <section data-v-4cf8372f="" class="response-evaluator__actions"><button data-v-4cf8372f="" data-testid="like-button" class="actions__button">
      <icon-stub data-v-4cf8372f="" next="false" icon="thumb_up" clickable="false" size="sm" scheme="neutral-cloudy"></icon-stub>I liked
    </button><button data-v-4cf8372f="" data-testid="dislike-button" class="actions__button">
      <icon-stub data-v-4cf8372f="" filled="false" next="false" icon="thumb_down" clickable="false" size="sm" scheme="neutral-cloudy"></icon-stub>I didn't like
    </button></section>
  <!--v-if-->
</section>"
`;
